Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2608,6 +2608,41 @@ def visit_CArgDeclNode(self, node):
         pass
 
 
+
+class IsCoroutineTransform(CythonTransform):
+    def visit_FuncDefNode(self, node):
+        if isinstance(node, (Nodes.AsyncDefNode, Nodes.IterableAsyncDefNode, Nodes.AsyncGenNode)):
```

Comment:
Note that you can intercept on these directly by naming convention, rather than giving the method a more general name and checking the type here.

But, rather than implementing this as a transform, I think having it in C would be much better. The particular object can just be an additional parameter for the `__Pyx_CyFunction_New()` C function, and the import can happen in a utility function. Or, the import could even happen on first access to the attribute (property) if it's still `NULL`, and setting it to `None` initially would make it a normal non-async function. Or something like that. In any case, the place where the function object is being created in C seems a good place to make the distinction and pass it into the `CyFunction_New()`, as a flag, object, or whatever.